netty/src/main/java/io/grpc/netty/AbstractNettyHandler.java

```diff
@@ -108,4 +123,117 @@ private void sendInitialConnectionWindow() throws Http2Exception {
       ctx.flush();
     }
   }
+
+  @VisibleForTesting
+  FlowControlPinger flowControlPing() {
+    return flowControlPing;
+  }
+
+  @VisibleForTesting
+  void setAutoTuneFlowControl(boolean isOn) {
+    autoTuneFlowControlOn = isOn;
+  }
+
+  /**
+   * Class for handling flow control pinging and flow control window updates as necessary.
+   */
+  final class FlowControlPinger {
+
+    private static final int MAX_WINDOW_SIZE = 8 * 1024 * 1024;
+    private int pingCount;
+    private int pingReturn;
+    private boolean pinging;
+    private int dataSizeSincePing;
+    private float lastBandwidth; // bytes per second
+    private long lastPingTime;
+
+    public int payload() {
+      return BDP_MEASUREMENT_PING;
+    }
+
+    public int maxWindow() {
+      return MAX_WINDOW_SIZE;
+    }
+
+    public void onDataRead(int dataLength, int paddingLength) {
+      if (!autoTuneFlowControlOn) {
+        return;
+      }
+      if (!isPinging()) {
+        setPinging(true);
+        sendPing(ctx());
+      }
+      incrementDataSincePing(dataLength + paddingLength);
+    }
+
+    public void updateWindow() throws Http2Exception {
+      if (!autoTuneFlowControlOn) {
+        return;
+      }
+      pingReturn++;
+      long elapsedTime = (System.nanoTime() - lastPingTime);
+      if (elapsedTime == 0) {
+        elapsedTime = 1;
+      }
+      long bandwidth = 0;
+      if (elapsedTime > 0) {
+        bandwidth = (getDataSincePing() * TimeUnit.SECONDS.toNanos(1)) / elapsedTime;
+      }
+      Http2LocalFlowController fc = decoder().flowController();
+      // Calculate new window size by doubling the observed BDP, but cap at max window
+      int targetWindow = Math.min(getDataSincePing() * 2, MAX_WINDOW_SIZE);
+      setPinging(false);
+      int currentWindow = fc.initialWindowSize(connection().connectionStream());
+      if (targetWindow > currentWindow && bandwidth > lastBandwidth) {
+        lastBandwidth = bandwidth;
+        int increase = targetWindow - currentWindow;
+        fc.incrementWindowSize(connection().connectionStream(), increase);
+        fc.initialWindowSize(targetWindow);
+        Http2Settings settings = new Http2Settings();
+        settings.initialWindowSize(targetWindow);
+        frameWriter().writeSettings(ctx(), settings, ctx().newPromise());
+      }
+
+    }
+
+    private boolean isPinging() {
+      return pinging;
+    }
+
+    private void setPinging(boolean pingOut) {
+      pinging = pingOut;
+    }
+
+    private void sendPing(ChannelHandlerContext ctx) {
+      setDataSizeSincePing(0);
```

Comment:
I only ever call sendPing from onDataRead which checks for pinging there, but if you think it makes more sense to check within to sendPing I'll move it.